File: Data Structures & Algorithms/subsets/submission-3.py
# sys.setrecursionlimit(3000)  

import copy
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        output = []
        
        def dfs(base, idx):
            nonlocal output
            
            if idx >= len(nums): return
            base.append(nums[idx])
            output.append(copy.deepcopy(base))
            dfs(base, idx+1)
            base.pop()
            dfs(base, idx+1)

        output.append([])
        dfs([], 0)
        return output
            

File: Data Structures & Algorithms/sum-of-all-subset-xor-totals/submission-3.cpp
class Solution {
    int dfs(const vector<int>& nums, int buf, int idx) {
        if (idx >= nums.size()) {
            return buf;
        }

        int acc = 0;
        acc += dfs(nums, buf^nums[idx], idx+1);
        acc += dfs(nums, buf, idx+1);
        return acc;
        
    }

public:
    int subsetXORSum(vector<int>& nums) {

        return dfs(nums, 0, 0);
        
    }
};